addons/maintainence-agent/maintenance_mode_controller.go

```diff
@@ -83,37 +72,59 @@ func (r *MaintenanceModeReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 	}
 
 	if !mmode.GetDeletionTimestamp().IsZero() || len(mmode.Spec.Modes) == 0 {
-		result, err := r.startMaintenanceActions(ctx, mmode, actionableCephClusters, false)
+		if !utils.ContainsString(mmode.GetFinalizers(), MaintenanceModeFinalizer) {
+			return ctrl.Result{}, nil
+		}
+
+		result, err := r.startMaintenanceActions(ctx, &mmode, actionableCephClusters, false)
 		if err != nil {
 			return result, err
 		}
-		mmode.Finalizers = utils.RemoveString(mmode.Finalizers, MaintenanceModeFinalizer)
 
+		mmode.Finalizers = utils.RemoveString(mmode.Finalizers, MaintenanceModeFinalizer)
 		if err := r.SpokeClient.Update(ctx, &mmode); err != nil {
 			klog.Error(err, "failed to remove finalizer from MaintenanceMode ", err)
 			return ctrl.Result{}, err
 		}
-		klog.Info("MaintenanceMode deleted, skipping reconciliation")
+
+		klog.Info("MaintenanceMode disabled")
 		return ctrl.Result{}, nil
-	} else {
-		_, err := r.startMaintenanceActions(ctx, mmode, actionableCephClusters, true)
-		if err != nil {
+	}
+
+	if !utils.ContainsString(mmode.GetFinalizers(), MaintenanceModeFinalizer) {
+		klog.Info("finalizer not found on MaintenanceMode. adding Finalizer ", MaintenanceModeFinalizer)
+		mmode.Finalizers = append(mmode.Finalizers, MaintenanceModeFinalizer)
+		if err := r.SpokeClient.Update(ctx, &mmode); err != nil {
+			klog.Error(err, "failed to add finalizer to MaintenanceMode", mmode.Name)
 			return ctrl.Result{}, err
 		}
 	}
 
-	return ctrl.Result{}, nil
+	if mmode.Status.State == ramenv1alpha1.MModeStateCompleted && mmode.Status.ObservedGeneration == mmode.Generation {
+		// All done for the current generation, return success
+		return ctrl.Result{}, nil
+	}
+
+	_, err = r.startMaintenanceActions(ctx, &mmode, actionableCephClusters, true)
+	statusErr := r.SpokeClient.Status().Update(ctx, &mmode)
```

Comment:
Update API status once at the end, else each update would cause the subsequent update to error out with a conflict. This prevented a delete from succeeding, and also prevented the MMode from reporting a state of Completed ever, as status updates constantly failed.